### ISSUE TITLE
Constant pool for class generated.ResultSetProcessorFactoryProvider...$SelectExprProcessorImpl has grown past JVM limit of 0xFFFF.

### ISSUE THREAD

**Issue by Rsverma** [espertechinc/esper#156]
Hi,

We have upgraded Esper from version 5.4.0 to 8.0, made required code changes related to common, compiler and runtime mentioned in "Information about esper version 8.0" articles.

We are facing an issue while compiling one of our existing EPL files.

org.codehaus.commons.compiler.CompileException: Line 55, Column 13: Constant pool for class generated.ResultSetProcessorFactoryProvider_<epl file name>$SelectExprProcessorImpl has grown past JVM limit of 0xFFFF.

The EPL code format:

```
@Priority(10)
@Name('RowCalculation')
@Description('Calculates row level data.')
insert into RowCalculation
Select 
	<Around 130 columns >
from 
	<Join on 15 windows>
```

  
Around 70 columns are of the format:

coalesce(exprCalculatefield1(S,T,O,A,FX,L,U,US,UA),0) as field1,

If we provide a constant value instead of calculating these columns, the code runs fine.
Please provide a way to optimize/spread the calculations so as to compile the code successfully.

[Exception from eclipse.txt](https://github.com/espertechinc/esper/files/3102677/Exception.from.eclipse.txt)

**Comment by raguessner** [comment]
I wasn't able to reproduce. Can you attach a test class please? my attempt is below.

    private static class ClientCompileLargeStmtManyUDFCalls implements RegressionExecution {
        public void run(RegressionEnvironment env) {
            StringBuilder buf = new StringBuilder();
            buf.append("select ");
            String delimiter = "";
            for (int i = 0; i < 1000; i++) {
                buf.append(delimiter);
                buf.append("coalesce(myFunction(theString, intPrimitive, intBoxed, doublePrimitive, doubleBoxed, floatPrimitive, floatBoxed, longPrimitive, longBoxed), 0) as ");
                buf.append("c" + i);
                delimiter = ",";
            }
            buf.append(" from SupportBean");
            EPCompiled compiled = env.compile(buf.toString());
        }
    }

**Comment by raguessner** [comment]
Or can you provide the generated code? This can be found in the log file right with the exception.

**Comment by Rsverma** [comment]
The issue does not occur when calling the same function again and again.
Try something like this:
buf.append("coalesce(myFunction"+i+"(theString, intPrimitive, intBoxed, doublePrimitive, doubleBoxed, floatPrimitive, floatBoxed, longPrimitive, longBoxed), 0) as ");

**Comment by raguessner** [comment]
Tried that, could not reproduce. Attach a test class.

**Comment by Rsverma** [comment]
Below is the Configuration and compile Deploy methods used:

`private static void configureCEPEngine(String configFile) throws Exception {
		try {

			File configFileAbs = new File(configFile);

			Configuration configuration = new Configuration();
			configuration.getRuntime().getExecution().setPrioritized(true);
			configuration.getRuntime().getThreading().setInternalTimerEnabled(false);
			configuration.getCompiler().getByteCode().setBusModifierEventType(EventTypeBusModifier.BUS);
			configuration.getCompiler().getByteCode().setAccessModifiersPublic();
			configuration.getCompiler().getByteCode().setAllowSubscriber(true);
						
			configuration.configure(configFileAbs);
			_epruntime= EPRuntimeProvider.getDefaultRuntime(configuration);
			_epruntime.initialize();
		} catch (Exception ex) {
			//Logging
			throw ex;
		}

}

private static void compileDeploy(Module epl) throws Exception {
		try {
		     // Build compiler arguments
		     CompilerArguments args = new CompilerArguments(_epruntime.getConfigurationDeepCopy());
		        	
		     // Make the existing EPL objects available to the compiler
			args.getPath().add(_epruntime.getRuntimePath());
					
			// Compile
			EPCompiled compiled = EPCompilerProvider.getCompiler().compile(epl, args);
		        	
		     EPDeployment deployment = _epruntime.getDeploymentService().deploy(compiled);
					
		     //Adding Listener
			for (EPStatement ep : deployment.getStatements()) {     
				
				if (ep != null)
					{
					_statementDict.put(ep.getName(), ep.getDeploymentId());
					}
			}	   
		} catch (Exception ex) {
		//Logging
		throw ex;
	}

}`

**Comment by Rsverma** [comment]
As we are deploying multiple modules dependant on one another it is slightly difficult to provide EPL code which compiles without its referenced files.

**Comment by raguessner** [comment]
I suggest to obtain the compiler output. The information is in http://esper.espertech.com/release-8.1.0/reference-esper/html_single/index.html#configuration-compiler-logging-code

**Comment by Rsverma** [comment]
Even after enabling the config entry, all I can find in the log file is the EPL text.
I don't think the byte code is being generated, as the code is failing at compile statement not at deploy.

**Comment by raguessner** [comment]
It is generated. You can alternatively get the code. The information is in http://esper.espertech.com/release-8.1.0/reference-esper/html_single/index.html#apicompiler-debugging

**Comment by Rsverma** [comment]
[ResultSetProcessorFactoryProvider_eplModules46rowCalculation46RowCalculationBase_RowCalculation.zip](https://github.com/espertechinc/esper/files/3231218/ResultSetProcessorFactoryProvider_eplModules46rowCalculation46RowCalculationBase_RowCalculation.zip)

**Comment by raguessner** [comment]
Can you compile setting "configuration.getCompiler().getExecution().isEnabledDeclaredExprValueCache()" to false?

**Comment by raguessner** [comment]
I don't see anywhere that there is a large number of constants in the zip file that you uploaded. Can you please provide a test class and the EPL.

**Comment by bernhardttom** [comment]
"javap -c -verbose X.class" shows the constant pool

**Comment by raguessner** [comment]
I'm reproducing it now.
        public void run(RegressionEnvironment env) {
            StringBuilder buf = new StringBuilder();
            buf.append("select ");
            String delimiter = "";
            for (int i = 0; i < 1000; i++) {
                buf.append(delimiter);
                buf.append("func(");
                String delimiterFunc = "";
                for (int j = 0; j < 40; j++) {
                    buf.append(delimiterFunc);
                    buf.append("'" + i + j).append("'");
                    delimiterFunc = ",";
                }
                buf.append(") as z" + i);
                delimiter = ",";
            }
            buf.append(" from SupportBean");

            EPCompiled compiled = env.compile(buf.toString());
        }

**Comment by raguessner** [comment]
Below reproduces.
```
       public void run(RegressionEnvironment env) {
            StringBuilder buf = new StringBuilder();
            buf.append("select ");
            String delimiter = "";
            for (int i = 0; i < 1000; i++) {
                buf.append(delimiter);
                buf.append("func(");
                String delimiterFunc = "";
                for (int j = 0; j < 40; j++) {
                    buf.append(delimiterFunc);
                    buf.append("'" + i + j).append("'");
                    delimiterFunc = ",";
                }
                buf.append(") as z" + i);
                delimiter = ",";
            }
            buf.append(" from SupportBean");

            EPCompiled compiled = env.compile(buf.toString());
        }
```

**Comment by raguessner** [comment]
Improvement to reproducing it is here. Instead of strings in the constant pool it is methods.
```
        public void run(RegressionEnvironment env) {
            StringBuilder buf = new StringBuilder();
            buf.append("select ");
            String delimiter = "";
            for (int i = 0; i < 1000; i++) {
                buf.append(delimiter);
                buf.append("((((((((((((((((((((((((1+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1)+1");
                buf.append(" as z" + i);
                delimiter = ",";
            }
            buf.append(" from SupportBean");

            EPCompiled compiled = env.compile(buf.toString());
        }

```

**Comment by Rsverma** [comment]
Thanks for the effort.
Is there any solution or workaround to this?

**Comment by raguessner** [comment]
Simplify the EPL- would be one way to work around. You can try the current build. The next release is soon.

**Comment by bernhardttom** [comment]
fix is part of 8.2 release
